### PR TITLE
Added parameter documentation on hover from function docString

### PR DESCRIPTION
**Pull request by drbrady8800**
closes [microsoft/pylance-release#2143](https://github.com/microsoft/pylance-release/issues/2143)

Screenshots:
Without docstring:
image

With docstring:
image

### CODE REVIEW

**Comment by drbrady8800** [comment]
> @microsoft-github-policy-service agree

@microsoft-github-policy-service agree

File: packages/pyright-internal/src/analyzer/binder.ts [hunk 1]
```diff
@@ -521,6 +522,7 @@ export class Binder extends ParseTreeWalker {
                 node.parameters.forEach((paramNode) => {
                     if (paramNode.name) {
                         const symbol = this._bindNameToScope(this._currentScope, paramNode.name);
+                        const docString = extractParameterDocumentation(ParseTreeUtils.getDocString(node?.suite?.statements ?? []) ?? "", paramNode.name.value) ?? undefined;
```
Comment (erictraut) [inline packages/pyright-internal/src/analyzer/binder.ts:525]: This line of code is difficult to understand with all of the `?` and `??` operators. I recommend adding `if` statements to make it more understandable.

**Review by erictraut — COMMENTED** [review]
I think this is a worthwhile change. One concern is that parameter symbols can be reassigned within a function (as long as the assigned value doesn't violate the type annotation), so displaying the parameter docstring for all instances of the symbol might be somewhat confusing. I wonder if it would be better to display the docstring only when hovering over the parameter itself?

**Comment by drbrady8800** [comment]
@erictraut Thank you for the feedback! To follow up on your question, the way I interpreted the discussion linked above was that the parameter's docstring would be displayed throughout the function. While the symbol can be reassigned I feel as though the docstring would still hold up in most cases. I agree that it is important to weigh the value that comes with displaying the docstring throughout (especially in longer / more complex functions) vs. the confusion that could arise if reassigned. I personally lean more toward the former, but a discussion on this would be great.

**Comment by erictraut** [comment]
@debonte, please discuss with pylance team and review the proposed change. I'm good with it if you are.

**Comment by debonte** [comment]
@erictraut, Pylance is good with this change.

File: packages/pyright-internal/src/analyzer/declaration.ts [hunk 2]
```diff
@@ -103,6 +103,10 @@ export interface ParameterDeclaration extends DeclarationBase {
     type: DeclarationType.Parameter;
     node: ParameterNode;
 
+    // documentation specified in the function's docstring (if any) can be
+    // associated with the parameter.
```
[the remark this answers is not in the document]
Comment (ifuatgucluer) [reply packages/pyright-internal/src/analyzer/declaration.ts]: This line of code is difficult to understand for me at now.

**Pull request merged**